uc-crux-llvm/src/UCCrux/LLVM/Bug/UndefinedBehaviorTag.hs

```diff
@@ -0,0 +1,630 @@
+{-
+Module           : UCCrux.LLVM.Bug.UndefinedBehaviorTag
+Description      : Representation of undefined behavior
+Copyright        : (c) Galois, Inc 2021
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+-}
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PolyKinds #-}
+
+module UCCrux.LLVM.Bug.UndefinedBehaviorTag
+  ( UndefinedBehaviorTag,
+    getUndefinedBehaviorTag,
+    makeUndefinedBehaviorTag,
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import           Data.Functor.Const (Const(Const))
+
+import           Data.Parameterized.TraversableF (fmapF)
+
+import           Lang.Crucible.LLVM.Errors.UndefinedBehavior (UndefinedBehavior)
+import qualified Lang.Crucible.LLVM.Errors.UndefinedBehavior as UB
+{- ORMOLU_ENABLE -}
+
+-- | A simplification of 'UndefinedBehavior' that keeps less information around.
+-- Used for deduplicating reports about possible bugs/errors in programs and
+-- explaining the provenance of inferred function preconditions.
+newtype UndefinedBehaviorTag =
+  UndefinedBehaviorTag { getUndefinedBehaviorTag :: UndefinedBehavior (Const ()) }
+
+makeUndefinedBehaviorTag :: UndefinedBehavior e -> UndefinedBehaviorTag
+makeUndefinedBehaviorTag = UndefinedBehaviorTag . fmapF (const (Const ()))
+
+-- | This instance is a coarsening of that for 'UndefinedBehavior'. In
+-- particular, there may be instances of 'UndefinedBehavior' that do not compare
+-- equal, but their projections under 'makeUndefinedBehaviorTag' do compare
+-- equal.
+--
+-- The less coarse this becomes, the more effortful it gets to write, but also
+-- the better it will be for deduplicating possible bugs/errors.
+instance Eq UndefinedBehaviorTag where
```

Comment:
Can this just call `testEqualityC`?

Reply:
That would be nice, but I tried it and couldn't make it work. Specializing the type of `testEqualityC`...
```haskell
(forall x y. f x -> f y -> Maybe (x :~: y)) -> t f -> t f -> Bool

(forall x y. f x -> f y -> Maybe (x :~: y)) -> 
  UndefinedBehavior f -> 
  UndefinedBehavior f -> 
  Bool

(forall x y. Const () x -> Const () y -> Maybe (x :~: y)) -> 
  UndefinedBehavior (Const ()) -> 
  UndefinedBehavior (Const ()) -> 
  Bool
```
There's the rub: There's no safe way to make a function that could produce `x :~: y` out of two `Const ()`s.

We could consider adding enough type information to `UndefinedBehavior` (i.e., adding a bunch of `NatRepr`s) so that we could make `UndefinedBehaviorTag ~ UndefinedBehavior TypeRepr`, which would be easier to handle.

Reply:
Ah, I just had another hacky thought. You can define:
```
data NoValues (tp :: CrucibleType) where
  NoValues :: NoValues BoolType

equalWithoutValue :: NoValues tp1 -> NoValues tp2 -> Maybe (tp1 :~: tp2)
equalWithoutValue NoValues NoValues = Just Refl
```

Reply:
Those are both good ideas, but I don't think either can work. Addressing the second one first:
```haskell
data OneValue (tp :: CrucibleType) where
  OneValue :: OneValue BoolType

equalWithoutValue :: OneValue tp1 -> OneValue tp2 -> Maybe (tp1 :~: tp2)
equalWithoutValue OneValue OneValue = Just Refl

newtype UndefinedBehaviorTag =
  UndefinedBehaviorTag { getUndefinedBehaviorTag :: UndefinedBehavior OneValue }

makeUndefinedBehaviorTag :: UndefinedBehavior e -> UndefinedBehaviorTag
makeUndefinedBehaviorTag = UndefinedBehaviorTag . fmapF (const OneValue)
```
yields
```
src/UCCrux/LLVM/Bug/UndefinedBehaviorTag.hs:46:58-71: error:
    • Couldn't match type ‘x’ with ‘BoolType’
      ‘x’ is a rigid type variable bound by
        a type expected by the context:
          forall (x :: CrucibleType). e x -> OneValue x
        at src/UCCrux/LLVM/Bug/UndefinedBehaviorTag.hs:46:57-72
      Expected type: e x -> OneValue x
        Actual type: e x -> OneValue BoolType
    • In the first argument of ‘fmapF’, namely ‘(const OneValue)’
      In the second argument of ‘(.)’, namely ‘fmapF (const OneValue)’
      In the expression: UndefinedBehaviorTag . fmapF (const OneValue)
   |
46 | makeUndefinedBehaviorTag = UndefinedBehaviorTag . fmapF (const OneValue)
   |
```
which makes sense - there are constructors of `UndefinedBehavior` like
```haskell
  FreeBadOffset ::
    (1 <= w) =>
    e (LLVMPointerType w) ->
    UndefinedBehavior e
```
but we have nothing to put in that field - there aren't any matching `OneValue` constructors that can have that type parameter (`LLVMPointerType w`).

And for the first one, it's also impossible to write `makeUndefinedBehaviorTag` - there just isn't enough information contained in an `UndefinedBehavior` to reconstruct a `TypeRepr` (which you can also see from the `FreeBadOffset` constructor above - it has a constraint on `w` but no `NatRepr w`). That's why I suggested:

> We could consider adding enough type information to UndefinedBehavior (i.e., adding a bunch of NatReprs) so that we could make UndefinedBehaviorTag ~ UndefinedBehavior TypeRepr, which would be easier to handle.

Reply:
I've pushed a version that uses `unsafeCoerce`.

Reply:
Hold on now... not happy with a solution that ends with "use unsafeCoerce" unless we really have no options.

Reply:
It looks like you can arrange to have an `IsExpr` contraint in the right places, so you should be able to get a `BaseTypeRepr` in the places where you need them.

Reply:
That right place might just be in `makeUndefinedBehaviorTag` - you could then use that constraint when mapping over terms to case on a type repr and produce a type-correct surrogate value

Reply:
And that surrogate value could just be the type repr

Reply:
Nice suggestion, @robdockins! That works like a charm. I did have to add
```haskell
llvmPointerType :: IsExpr (SymExpr sym) => LLVMPtr sym w -> TypeRepr (LLVMPointerType w)
llvmPointerType ptr =
  case exprType (llvmPointerOffset ptr) of
    BaseBVRepr w -> LLVMPointerRepr w
```
to `Lang.Crucible.LLVM.MemModel.Pointer` to make it work, but that seems OK to me?